Create set_cell_width() function

trhille opened the pull request:
Create a function to determine cell_width based on input parameters in the config file. This currently contains three different cell spacing functions based on log10(speed), distance to ice edge, and distance to the grounding line that can be turned on or off in the config file. Minimum and maximum cell width and the distances over which cell width is varied are also set in the config file. Implement a use case for Humboldt and update User's Guide. This will be used by all mesh-generating land ice test cases based on real-world geometries.

Note: The mesh created here is not bit-for-bit with the previous version. It is not clear why this is the case, but the differences are trivial.

pep8speaks commented:
Hello @trhille! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`compass/landice/mesh.py`](https://github.com/MPAS-Dev/compass/blob/61de4d0b2c9806963b8f724a8e0fcb4c5f53cff8/compass/landice/mesh.py):

> [Line 177:30](https://github.com/MPAS-Dev/compass/blob/61de4d0b2c9806963b8f724a8e0fcb4c5f53cff8/compass/landice/mesh.py#L177): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 187:30](https://github.com/MPAS-Dev/compass/blob/61de4d0b2c9806963b8f724a8e0fcb4c5f53cff8/compass/landice/mesh.py#L187): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator

* In the file [`compass/landice/tests/humboldt/mesh.py`](https://github.com/MPAS-Dev/compass/blob/61de4d0b2c9806963b8f724a8e0fcb4c5f53cff8/compass/landice/tests/humboldt/mesh.py):

> [Line 274:1](https://github.com/MPAS-Dev/compass/blob/61de4d0b2c9806963b8f724a8e0fcb4c5f53cff8/compass/landice/tests/humboldt/mesh.py#L274): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 275:9](https://github.com/MPAS-Dev/compass/blob/61de4d0b2c9806963b8f724a8e0fcb4c5f53cff8/compass/landice/tests/humboldt/mesh.py#L275): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,100 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+    thk : numpy.ndarray
+        Ice thickness field from gridded dataset,
+        usually after trimming to flood fill mask
+    vx : numpy.ndarray
```
```suggestion
    vx : numpy.ndarray, optional
```

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,100 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+        x-component of ice velocity from gridded dataset,
+        usually after trimming to flood fill mask. Can be set to None
+        if useSpeed == 'False' in config file.
+    vy : numpy.ndarray
```
```suggestion
    vy : numpy.ndarray, optional
```

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,100 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+        y-component of ice velocity from gridded dataset,
+        usually after trimming to flood fill mask. Can be set to None
+        if useSpeed == 'False' in config file.
+    distToEdge : numpy.ndarray
```
I'd prefer to move away from camel case in the framework as much as possible, but particularly in function arguments and config options (see below).
```suggestion
    dist_to_edge : numpy.ndarray, optional
```

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,100 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+        Can be set to None if useDistToEdge == 'False' in config file and you
+        do not want to set large cell_width where cells will be culled anyway,
+        but this is not recommended.
+    distToGroundingLine : numpy.ndarray
```
```suggestion
    dist_to_grounding_line : numpy.ndarray, optional
```

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,100 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+        usually after trimming to flood fill mask
+    vx : numpy.ndarray
+        x-component of ice velocity from gridded dataset,
+        usually after trimming to flood fill mask. Can be set to None
+        if useSpeed == 'False' in config file.
```
I'd suggest putting some things in double back quotes so they come out looking like code in the docs:
```suggestion
        usually after trimming to flood fill mask. Can be set to ``None``
        if ``useSpeed == False`` in config file.
```
and the same below.

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,100 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+
+    Parameters
+    ----------
+    section : str
+        Section of the config file from which to read parameters
```
Either here or above in the main text of the docstring, you will need to specify all of the config options that are required to be in the given config section for the function to work.

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,100 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+                            right=minSpac)
+        spacing[thk == 0.0] = minSpac
+    else:
+        spacing = thk * 0. + maxSpac
```
```suggestion
        spacing = maxSpac*np.ones_like(thk)
```

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,100 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+                             right=maxSpac)
+        spacing2[thk == 0.0] = minSpac
+    else:
+        spacing2 = thk * 0. + maxSpac
```
```suggestion
        spacing2 = maxSpac*np.ones_like(thk)
```

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,100 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+                             right=maxSpac)
+        spacing3[thk == 0.0] = minSpac
+    else:
+        spacing3 = thk * 0. + maxSpac
```
```suggestion
        spacing3 = maxSpac*np.ones_like(thk)
```

xylar commented on docs/users_guide/landice/test_groups/humboldt.rst:
```diff
@@ -28,6 +28,25 @@ The test group uses the following default config options:
     # to cull based on distance from margin.
     cullDistance = 5.0
 
+    # mesh density parameters
+    # minimum cell spacing (meters)
+    minSpac = 1.e3
+    # maximum cell spacing (meters)
+    maxSpac = 1.e4
+    # log10 of max speed (m/yr) for cell spacing
+    highLogSpeed = 2.5
+    # log10 of min speed (m/yr) for cell spacing
+    lowLogSpeed = 0.75
+    # distance at which cell spacing = maxSpac (meters)
+    highDist = 1.e5
+    # distance within which cell spacing = minSpac (meters)
+    lowDist = 1.e4
+    
+    # mesh density functions
+    useSpeed = True
+    useDistToGroundingLine = False
+    useDistToEdge = True
```
Config options end up in all lowercase after they get parsed and written out again as a new file.  That makes camel case particularly hard to deal with.  Could you convert these to be lowercase with underscores?  It would be clearer if you convert the corresponding variables in the function to the same naming convention (which is also what python prefers) but I don't feel as strongly about that.

xylar replied:
Make sure the docstring gets updated with the corresponding changes.

xylar reviewed (changes requested):
This is really great!  I have a several suggested formatting changes to clean things up.

trhille commented:
@xylar thanks for your review. I addressed these issues in commits [cdc6557](https://github.com/MPAS-Dev/compass/pull/321/commits/cdc655765a951eeb43de7b4dc4eab719ac5e7506) and [c47701b](https://github.com/MPAS-Dev/compass/pull/321/commits/c47701b96faf56bb7d3b0bad4975db706f3f7482). The PEP8 issues currently listed above don't make any sense, however, as none of the lines referred to actually contain those issues. I'm not sure what went wrong there.

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,103 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+    Set cell widths based on settings in config file to pass to
+    :py:func:`mpas_tools.mesh.creation.build_mesh.build_planar_mesh()`.
+    Requires the following options to be set in the given config section:
+    ``min_spac``, ``max_spac``, ``high_log_speed``, ``low_log_speed``,
+    ``high_dist``, ``low_dist``, and ``cull_distance``.
```
It looks like `use_speed`, `use_dist_to_grounding_line` and `use_dist_to_edge` are also required, right?

trhille replied:
Oops, yes. Apologies for sloppiness. This and suggestions below are addressed in https://github.com/MPAS-Dev/compass/pull/321/commits/ff7525d211d68e13a93f233841235422f0b35bce.

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,103 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+    vx : numpy.ndarray, optional
+        x-component of ice velocity from gridded dataset,
+        usually after trimming to flood fill mask. Can be set to ``None``
+        if ``useSpeed == False`` in config file.
```
```suggestion
        if ``use_speed == False`` in config file.
```

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,103 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+    vy : numpy.ndarray, optional
+        y-component of ice velocity from gridded dataset,
+        usually after trimming to flood fill mask. Can be set to ``None``
+        if ``useSpeed == False`` in config file.
```
```suggestion
        if ``use_speed == False`` in config file.
```

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,103 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+        if ``useSpeed == False`` in config file.
+    dist_to_edge : numpy.ndarray, optional
+        Distance from each cell to ice edge, calculated in separate function.
+        Can be set to ``None`` if ``useDistToEdge == False`` in config file
```
```suggestion
        Can be set to ``None`` if ``use_dist_to_ddge == False`` in config file
```

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,103 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+    dist_to_edge : numpy.ndarray, optional
+        Distance from each cell to ice edge, calculated in separate function.
+        Can be set to ``None`` if ``useDistToEdge == False`` in config file
+        and you do not want to set large cell_width where cells will be culled
```
```suggestion
        and you do not want to set large ``cell_width`` where cells will be culled
```

xylar commented on compass/landice/mesh.py:
```diff
@@ -102,3 +102,103 @@ def set_rectangular_geom_points_and_edges(xmin, xmax, ymin, ymax):
+    dist_to_grounding_line : numpy.ndarray, optional
+        Distance from each cell to grounding line, calculated in separate
+        function.  Can be set to ``None`` if
+        ``useDistToGroundingLine == False`` in config file.
```
```suggestion
        ``use_dist_to_grounding_line == False`` in config file.
```

xylar reviewed (changes requested):
Thanks for the changes!  A few more small things.  Almost there...

xylar reviewed (approved):
Looks great to me!

Probably run one final test to make sure things haven't somehow broken in these updates.  But that can wait until @matthewhoffman has reviewed.

matthewhoffman reviewed (approved):
I could potentially see wanting different transition distances for distance to GL and distance to edge, but I don't see a reason to over engineer this now.  This looks great, and we can always make adjustments later as specific use cases actually come up.

matthewhoffman commented:
@trhille , let me know if your final tests pass, and then I can merge this.

trhille commented:
@matthewhoffman, the final test passed. Thanks!

trhille commented:
> I could potentially see wanting different transition distances for distance to GL and distance to edge, but I don't see a reason to over engineer this now. This looks great, and we can always make adjustments later as specific use cases actually come up.

Agreed

trhille commented:
Sorry, hit the 'Close with comment' button by accident.
